pytket/extensions/qiskit/backends/crosstalk_model.py

```diff
@@ -0,0 +1,378 @@
+        if slice_idx == n_slices:
+            self._slices.append([inst])
+        else:
+            self._slices[slice_idx].append(inst)
+        # update frontier
+        for q in args:
+            frontier[q] = slice_idx + 1
+
+    def _fill_gaps(self, frontier: Dict[Qubit, int]) -> None:
+        """Fill the gaps in the slices with identity `Unitary1qBox`es"""
+        for idx, s in enumerate(self._slices):
+            slice_qubits = set().union(*[self._get_qubits(inst) for inst in s])
+            gap_qs = self.all_qubits - slice_qubits
+            for q in gap_qs:
+                # only fill up to the frontier
+                if idx < frontier[q]:
+                    s.append(Command(self.Ibox, [q]))
+
+    def sort_and_fill_gaps(self) -> None:
+        """Sort splices so each slice only contains independent instructions"""
+        old_slices = self._slices.copy()
+        self._slices = []
+        frontier = {q: 0 for q in self.all_qubits}
+        for s in old_slices:
+            self._append(s, frontier)
+        self._fill_gaps(frontier)
+
+    # Unitary factorisation
+    @staticmethod
+    def _matrix_power(u: np.ndarray, p: float) -> None:
+        """Raise a matrix to the power p via eigen decomp"""
+        values, vectors = np.linalg.eig(u)
+        gate = np.zeros(u.shape)
+        for i in range(len(values)):
+            gate = (
+                gate
+                + np.power(values[i] + 0j, p)
+                * vectors[:, [i]]
+                @ vectors[:, [i]].conj().T
+            )
+        return gate
+
+    @staticmethod
+    def _get_ubox(u: np.ndarray) -> Union[Unitary1qBox, Unitary2qBox, Unitary3qBox]:
+        """Return a UnitaryxqBox for a given unitary"""
+        if u.shape[0] == 2:
+            return Unitary1qBox(u)
+        if u.shape[0] == 4:
+            return Unitary2qBox(u)
+        if u.shape[0] == 8:
+            return Unitary3qBox(u)
+        raise ValueError(f"Unsupported unitary shape: {u.shape}")
+
+    def unitary_factorisation(self) -> None:
+        """For each unitary U with time D, factorise it into N*D unitaries u_i, such that
+        u_0;u1;...;u_(N*D-1) = U. Store the factorised circuit as a list of slices
+        """
+        for cmd in self.circ:
+            if cmd.op.type in [OpType.Measure, OpType.Reset]:
+                self._slices.append(cmd)
+            else:
+                gt = self.gate_times.get((cmd.op.type, tuple(cmd.qubits)))
+                if self.ct_params.virtual_z and cmd.op.type == OpType.Z:
+                    gt = 1 / self.N
+                if gt is None:
+                    raise ValueError(
+                        f"No gate time for OpType {cmd.op.type} on qubits {cmd.qubits}"
+                    )
+                u = cmd.op.get_unitary()
+                if (self.N * gt) % 1 > 1e-10:
+                    raise ValueError(
+                        f"Command {cmd} cannot be factorised into equal slices"
+                    )
```

Comment:
I'm developing a better understanding of N as I read :P . So N is (seconds per slice)^{-1}? In particular the units of N are s^{-1}?

Reply:
This impose that the gates times should all have 'common factors'. I wonder how often that actually happens with device characterisation information? I think this makes it more important to make customisable gates times easy to input.